### PR TITLE
Override W&B data on a resumed training

### LINKED ISSUE #594: weights and biases reporting doesn't work correctly when resuming training after a spot termination
Over in #580 we've been working on getting spot instances re-enabled. That part works correctly now, and training is continued from the last checkpoint, but I noticed in our latest run that reporting is turned off when resuming.

In [run #1](https://firefox-ci-tc.services.mozilla.com/tasks/U9n3MagKRfWfZgTf4Qu6fw/runs/1/logs/public/logs/live.log) we had:
```
[task 2024-05-13T23:07:22.900Z] wandb: Currently logged in as: moz-translations-wandb-bot (moz-translations). Use `wandb login --relogin` to force relogin
[task 2024-05-13T23:07:23.803Z] wandb: wandb version 0.17.0 is available!  To upgrade, please run:
[task 2024-05-13T23:07:23.803Z] wandb:  $ pip install wandb --upgrade
[task 2024-05-13T23:07:23.803Z] wandb: Tracking run with wandb version 0.16.1
[task 2024-05-13T23:07:23.803Z] wandb: Run data is saved locally in /home/ubuntu/tasks/task_171564135626786/checkouts/vcs/pipeline/train/wandb/run-20240513_230722-eeqsrpg6
[task 2024-05-13T23:07:23.803Z] wandb: Run `wandb offline` to turn off syncing.
[task 2024-05-13T23:07:23.805Z] wandb: Syncing run backwards
[task 2024-05-13T23:07:23.805Z] wandb: ⭐️ View project at https://wandb.ai/moz-translations/en-ru
[task 2024-05-13T23:07:23.805Z] wandb: 🚀 View run at https://wandb.ai/moz-translations/en-ru/runs/eeqsrpg6
```

But in [run #2](https://firefox-ci-tc.services.mozilla.com/tasks/U9n3MagKRfWfZgTf4Qu6fw/runs/2/logs/public/logs/live.log) we got:
```
[task 2024-05-14T00:51:23.329Z] [tracking WARNING] This run already exists on W&B: [<Run moz-translations/en-ru/eeqsrpg6 (running)>]. No data will be published.
```

Note that when resuming we usually will end up redoing some work that happens after the last checkpoint, but before the termination happened -- I'm not sure the best way to handle this in W&B - I imagine others will have a better idea.

### PR DESCRIPTION
**Pull request by vrigal**
Closes #594

### CODE REVIEW

File: tracking/translations_parser/publishers.py [hunk 1]
```diff
@@ -99,16 +99,25 @@ def open(self, parser) -> None:
...
                 )
-                if len(existing_runs) > 0:
+            if len(existing_runs) > 0:
+                if os.environ.get("RUN_ID", 0) > 0:
+                    logger.info(
+                        "Training has been resumed from an earlier run, data will be overwritten on W&B."
+                    )
+                    for run in existing_runs:
+                        logger.warning(f"Deleting existing run {run.display_name}.")
+                        run.delete()
```
Comment (eu9ene) [inline tracking/translations_parser/publishers.py:119]: @vrigal it's not supposed to delete the data from the previous run. It should continue tracking like the restart never happened.

**Review by eu9ene — CHANGES_REQUESTED** [review]

File: tracking/translations_parser/publishers.py (hunk 1 above)
Comment (eu9ene) [reply tracking/translations_parser/publishers.py:119]: See https://docs.wandb.ai/guides/runs/resuming#resuming-guidance

**Comment by vrigal** [comment]
I applied the changes (to resume an existing run on W&B) and tested here: https://wandb.ai/teklia/test/runs/tv2vtdha

> we can just pass the artifacts dir to the script parse_tc_logs --from-stream -v --wandb-artifacts ${model_dir}.

Its probably worth opening an issue for this. W&B seems to sync (override) logs & artifacts automatically. (https://wandb.ai/teklia/test/groups/test/files/output.log).

> One complication here is that we continue training from the last checkpoint so there will be small overlap with previously written data

W&B should handle this automatically, as it ignores data with a step inferior the last data written.
I think it should simply trigger some warnings from the W&B client.

**Comment by eu9ene** [comment]
@vrigal one thing I don't understand is who sets "RUN_ID" env?

**Review by eu9ene — APPROVED** [review]
Ok, this can work but from the code perspective there are some issues:
- w&b publisher should not know anything about Taskcluster environment variables
- why don't we just use resume="allow" so that if the run already exists, it continues automatically? I don't think there is a use case where we have a run with the same name when running things on Taskcluster except the restart of the same task. I guess this logic was implemented in the publisher for publishing offline experiments to prevent republishing the same ones. In this case, the publisher should accept an argument set by a cli that indicates what to do if the run already exists.
 
Other known issues:
- W&B drops overlapped data from the new run instead of overwriting (`wandb: WARNING (User provided step: 15000 is less than current step: 15001. Dropping entry: {'gnorm': 0.7161, '_timestamp': 1715790811.6654584}). `)

I think since it kind of works we can merge it to unblock enabling spot instances but we should address those issues later.

**Comment by vrigal** [comment]
> why don't we just use resume="allow" so that if the run already exists, it continues automatically? I don't think there is a use case where we have a run with the same name when running things on Taskcluster except the restart of the same task. I guess this logic was implemented in the publisher for publishing offline experiments to prevent republishing the same ones. In this case, the publisher should accept an argument set by a cli that indicates what to do if the run already exists.

This is an old issue. The simpler way to handle this would be to use `<group>-<model>` as a unique ID in W&B.
It should be possible to keep  resume="allow" then. It would be compatible with the override option and would probably work in most case (W&B drops overlapped data, as you mentioned), but requires some important changes in the code.

**Comment by eu9ene** [comment]
> > why don't we just use resume="allow" so that if the run already exists, it continues automatically? I don't think there is a use case where we have a run with the same name when running things on Taskcluster except the restart of the same task. I guess this logic was implemented in the publisher for publishing offline experiments to prevent republishing the same ones. In this case, the publisher should accept an argument set by a cli that indicates what to do if the run already exists.
> 
> This is an old issue. The simpler way to handle this would be to use `<group>-<model>` as a unique ID in W&B. It should be possible to keep resume="allow" then. It would be compatible with the override option and would probably work in most case (W&B drops overlapped data, as you mentioned), but requires some important changes in the code.

We can rethink all that in #408, but I would use UID in model names as a last resort because they would clutter the dashboards.

**Comment by vrigal** [comment]
@eu9ene to be clear, run ID (used to identify a run) is different that run name (used to display graphs). For now we do not use an ID, it is automatically set by W&B (e.g. `brmhnekj`) which guarantees unicity. But if there is a unique way to identify a run (e.g. `<group>-<model_name>`) it could help with deletion(override) or resuming a run. It can be a separate issue than #408. I'll write an issue for this.

**Pull request merged**